Preliminary support for modules/externals + CLI subcommands

AltGr opened the pull request:
- Module support is far from ready yet, but since it doesn't seem to break the
  existing in its current state, it may be worth merging already to avoid
  conflicts later on

- There are a few refactors that may also be worthwile (rename of `IdentName` to
  `Ident`, more uses of the AST mapper)

- Included, and maybe most important here is a refactor of command-line parsing
  to use sub-commands ; this allows different backends but also plugins to have
  specific flags. See 52b1c25d5 for details.

I can take the time to split the 3rd point into a separate PR if more discussion
is needed for either that or the rest, or for easier review if having all this
bundled is a problem.

denismerigoux commented on compiler/shared_ast/interpreter.ml:
```diff
@@ -375,10 +390,226 @@ let rec evaluate_operator
+(* let f e = (e : (< .. > as 'a, 't) gexpr :> (< custom : _; 'a; .. >, 't) gexpr ) 
+ * 
+ * let f (type a) ((e: (< custom: a; .. >, 't) naked_gexpr), t) =
+ *   if false then ECustom { obj = Obj.repr (); targs = []; tret = (TLit TUnit, Pos.no_pos) }, t
```
should this dead code be removed?

denismerigoux commented on compiler/shared_ast/interpreter.ml:
```diff
@@ -552,6 +804,142 @@ let rec evaluate_expr :
+ *       ; custom : 'm }
+ *       ->
+ *       (< monomorphic : 'a
+ *        ; polymorphic : 'b
```
dead code also?

denismerigoux commented on compiler/shared_ast/interpreter.ml:
```diff
@@ -552,6 +804,142 @@ let rec evaluate_expr :
+  in
+  Expr.unbox (f e)
+
+let evaluate_expr ctx e = delcustom (evaluate_expr ctx (addcustom e))
```
this shenanigan could gain from some explanations :)

denismerigoux commented on compiler/shared_ast/optimizations.ml:
```diff
@@ -184,7 +184,7 @@ let rec optimize_expr :
       Mark.remove (StructField.Map.find field fields)
     | EDefault { excepts; just; cons } -> (
-      (* TODO: mechanically prove each of these optimizations correct :) *)
+      (* TODO: mechanically prove each of these optimizations correct *)
```
:(

AltGr replied:
(my editor was'nt clever enough to pair the parens of the comment regardless of the emoji… now I feel bad for killing an innocent smile :cry: )

denismerigoux commented on compiler/shared_ast/qident.mli:
```diff
@@ -0,0 +1,36 @@
+    separate compilation units. *)
+
+type modname = string
+(** Expected to be a uident *)
```
That means you'll change these afterwards?

AltGr replied:
No, just that the string is assumed to start with an uppercase letter

denismerigoux reviewed (approved):
Thanks, I reviewed it all and it's fine to merge bundled like this!

AltGr commented on compiler/desugared/from_surface.ml:
```diff
@@ -1248,37 +1248,55 @@ let process_topdef
+  let expr_opt =
+    match def.S.topdef_expr, def.S.topdef_args with
+    | None, _ -> None
+    | Some e, None -> Some (Expr.unbox (translate_expr None None ctxt e))
```
Here we are dealing with top-level definitions (before program linearisation), so they are expected to be closed.

AltGr commented:
Ouch indeed I had left some dead bodies lying around :zombie: 
Now cleaned up and added some clarifications (plus a function `Expr.unbox_closed` that adds an assertion to `Expr.unbox`)
